Implement API v3 /path/data/ endpoint.

btimby opened the pull request:
`createWriteStream()` can accept an offset as an option. If provided, it will perform an HTTP PATCH request. This will append to the named file.

Added `deleteFile()` to use streamlined deletion of files (no operation / task).

Pushed error handling and retry handling up to `request()` method, so all operations benefit. Fixed metrics to include throttle retries in overall request time. Also standardized errors to always provide the request involved in the error (now that `request()` handles error callbacks).

Updated throttle header name (to `Retry-After`) for newer `django-rest-framework`. Fixed throttle handling to throw when a request cannot be automatically retried.

Removed `X-File-Name` workaround as API v3 no longer needs it, path is in URL.

Closes https://github.com/smartfile/client-node/issues/43

btimby commented:
`parseJson` in `upload` is (nearly) identical to a large portion of `requestJson`. Make `upload.parseJson` global and use it for both.

btimby commented:
Optionally, use `requestJson` for put and patch.

cabarnes commented on lib/fs/filesystem.js:
```diff
@@ -254,16 +270,40 @@ class FileSystem {
+
+      if (options.timestamp) {
+        const ts = moment.unix(options.timestamp);
+        options.headers['If-Unmodified-since'] = ts.format('ddd, d M YYYY HH:mm:ss GMT');
```
Any reason not to use ISO format?

btimby replied:
I was going by the spec of the `If-Unmodified-Since` header.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since

cabarnes replied:
Would this need a capital `S` then?

btimby replied:
Yes. Done.

btimby commented:
I added a handler for request errors. I found that without this, we get failed tests downstream. This was in the old `requestJSON` function and I did not "push" it up to request when I refactored. I also removed throttling handling for POST requests (this was broken).

btimby commented:
Verified these changes fix downstream tests. They failed because a 404 error (nock miss actually) would throw an error rather than returning it via callback where we handle it gracefully.
